### PR TITLE
chore: [CDS-79347]: Added Factory pattern to add the widgets and components for RPF with metadata

### PR DESCRIPTION
**Pull request by PratyushGar**
### Summary

image

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Review by srdjan-harness — APPROVED** [review]

File: src/modules/70-pipeline/y1/components/ComponentFactory/Widgets/TextComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import { InputsFormValues } from '../../InputsForm/InputsForm'
+
+function TextWidget(props: ComponentProps<InputsFormValues>): JSX.Element {
+  const { metadata, allowableTypes, readonly, path } = props
+  const { expressions } = useVariablesExpression()
+  const { NG_EXPRESSIONS_NEW_INPUT_ELEMENT } = useFeatureFlags()
+
+  return (
+    <FormInput.MultiTextInput
+      className="variableInput"
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/Widgets/TextComponent.tsx:23]: Is this className needed?

File: src/modules/70-pipeline/y1/components/ComponentFactory/Widget.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  factory: ComponentFactory
+  path: string
+  readonly?: boolean
+  formikRef?: ComponentFormikFowardRef<T>
+  customComponentProps?: U
+  allowableTypes: AllowedTypes
+  metadata: RuntimeInput
+}
+
+export function setFormikRef<T = unknown, U = unknown>(ref: ComponentFormikFowardRef<T>, formik: FormikProps<U>): void {
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/Widget.tsx:49]: Is this used in this PR?

File: src/modules/70-pipeline/y1/components/ComponentFactory/ComponentInterface.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  Connector = 'connector',
+  Email = 'email',
+  Timeout = 'timeout'
+}
+
+export enum HarnessDerivedComponentType {
+  ShellScript = 'shell_script'
+}
+
+export type ComponentType = HarnessPrimitiveComponentType | HarnessDerivedComponentType
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/ComponentInterface.ts:23]: Let's rename to InputComponentType

File: src/modules/70-pipeline/y1/components/ComponentFactory/Widget.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  if (!ref) return
+
+  if (typeof ref === 'function') {
+    return
+  }
+
+  ref.current = formik as unknown as FormikProps<T>
+}
+
+export abstract class Widget<T> {
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/Widget.tsx:59]:
```suggestion
export abstract class InputComponent<T> {
```

File: src/modules/70-pipeline/y1/components/ComponentFactory/ComponentWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  // internalAPIType: ComponentType
+  path?: string
+  onUpdate?: (data: unknown) => void
+  onChange?: (data: unknown) => void
+  readonly?: boolean
+  allowableTypes: AllowedTypes
+  children?: React.ReactNode
+}
+
+export function ComponentWidget<T = unknown, U = unknown>(
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/ComponentWidget.tsx:25]: InputComponentRenderer

File: src/modules/70-pipeline/y1/components/ComponentFactory/WidgetFactory.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+import { ComponentFactory } from './ComponentFactory'
+import { NumberComponent } from './Widgets/NumberComponent'
+import { TextAreaComponent } from './Widgets/TextAreaComponent'
+import { TextComponent } from './Widgets/TextComponent'
+
+const widgetFactory = new ComponentFactory()
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/WidgetFactory.ts:13]: An instance of `InputComponentFactory` can be just `inputComponentFactory`.

File: src/modules/70-pipeline/y1/components/ComponentFactory/Widgets/NumberComponent.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+          disabled: readonly,
+          type: 'number'
+        },
+        allowableTypes
+      }}
+    />
+  )
+}
+
+export class NumberComponent extends Widget<InputsFormValues> {
```
Comment (rameshx) [inline src/modules/70-pipeline/y1/components/ComponentFactory/Widgets/NumberComponent.tsx:41]: Let's call these `NumberInput`, `TextInput`, etc.

File: src/modules/70-pipeline/y1/components/ComponentFactory/Widget.tsx (hunk 2 above)
Comment (PratyushGar) [reply src/modules/70-pipeline/y1/components/ComponentFactory/Widget.tsx:49]: Not in this PR, but would be useful for our future widgets implementation on the edit view

**Review by rameshx — APPROVED** [review]

**Pull request merged**